modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java

```diff
@@ -238,21 +260,36 @@ private static List<Integer> mergeWithoutDuplicates(List<Integer> l0, List<Integ
     private void addListenersBusy() {
         catalogService.listen(INDEX_CREATE, fromConsumer(this::onIndexCreated));
         catalogService.listen(INDEX_REMOVED, fromConsumer(this::onIndexRemoved));
+
+        lowWatermark.addUpdateListener(this::onLwmChanged);
     }
 
     private void onIndexRemoved(RemoveIndexEventParameters parameters) {
         inBusyLock(busyLock, () -> {
             int indexId = parameters.indexId();
             int catalogVersion = parameters.catalogVersion();
 
-            CatalogIndexDescriptor index = indexBusy(indexId, catalogVersion - 1);
+            lowWatermark.getLowWatermarkSafe(lwm -> {
+                int lwmCatalogVersion = catalogService.activeCatalogVersion(hybridTimestampToLong(lwm));
 
-            if (index.status() == AVAILABLE) {
-                // On drop table event.
-                readOnlyIndexes.add(new ReadOnlyIndexInfo(index, catalogActivationTimestampBusy(catalogVersion)));
-            } else if (index.status() == STOPPING) {
-                readOnlyIndexes.add(new ReadOnlyIndexInfo(index, findStoppingActivationTsBusy(indexId, catalogVersion - 1)));
-            }
+                if (catalogVersion <= lwmCatalogVersion) {
```

Comment:
Here, an index is not tracked anymore if the moment of its **removal** is below the LWM. This seems correct, but we can stop tracking the index earlier: when it's **dropped**. If it's dropped with its table, it's the same as removal; but if it's dropped individually, then the moment of its drop is the moment when it becomes STOPPED.

Same thing applies to the recovery logic and the LWM listener.

I'm not sure if this needs to be fixed now. Maybe just create a ticket and add TODOs?

Reply:
Do I understand correctly that you mean that when the index goes to STOPPED and the low watermark has already occurred, then we can immediately drop the index?

Let me not leave TODO for now and just create a ticket for this?

Reply:
No, while an index is in `STOPPING`, it cannot be destroyed yet. But it can be destroyed if it's removed AND the moment of the switch to `STOPPING` (not of removal) is below LWM.

I'm fine with just filing a ticket.